shared/validate/validate.go

```diff
@@ -763,3 +763,46 @@ func IsAbsFilePath(value string) error {
 
 	return nil
 }
+
+func ParseNetworkVLANList(rawVLANValues []string) ([]string, error) {
+	var networkVLANList []string
+	for _, vlan := range rawVLANValues {
+		if err := IsNetworkVLAN(vlan); err != nil {
```

Comment:
The `ParseNetworkVLANRange` should be modified to be able to cope with a single VLAN ID (like https://github.com/lxc/lxd/blob/1972fb216a58530de901b0e4c7e4c2adeb85e2ab/shared/validate/validate.go#L93), so no need to call this separately here.

Reply:
The approach is a bit different here. Different then `ParseUint32Range`, the VLAN range needs to be expanded. Thus, this function serves the purpose to:

1. Check whether a VLAN input is a single VLAN or a range. 
2. Validate and parse the VLAN input into a list of consecutive VLAN IDs.

Reply:
Thus, I think it makes it easier for `ParseVLANRange` to just parse/cope with the VLAN range, and let `IsNetworkVLAN` takes care of the single VLAN IDs.

Reply:
I think something like this will work well:

1. Update `ParseNetworkVLANRange` to behave similar to `ParseUint32Range`. Handling both single and range values, and returning native values, and with comments.
```go
// ParseNetworkVLANRange parses a VLAN range in the form "number" or "start-end".
// Returns the start number and the size of the range.
func ParseNetworkVLANRange(value string) (uint, uint, error) {
	rangeParts := strings.SplitN(value, "-", 2)
	rangeLen := len(rangeParts)
	if rangeLen != 1 && rangeLen != 2 {
		return 0, 0, fmt.Errorf("VLAN range must contain a single number or start and end numbers")
	}

	err := IsNetworkVLAN(rangeParts[0])
	if err != nil {
		return 0, 0, err
	}

	startNum, err := strconv.Atoi(rangeParts[0])
	if err != nil {
		return 0, 0, fmt.Errorf("Invalid VLAN ID %q", value)
	}

	var rangeSize uint = 1

	if rangeLen == 2 {
		err := IsNetworkVLAN(rangeParts[1])
		if err != nil {
			return 0, 0, err
		}

		endNum, err := strconv.Atoi(rangeParts[1])
		if err != nil {
			return 0, 0, fmt.Errorf("Invalid end VLAN ID %q", value)
		}

		if startNum >= endNum {
			return 0, 0, fmt.Errorf("Start VLAN number %d must be lower than end number %d", startNum, endNum)
		}

		rangeSize += uint(endNum) - uint(startNum)
	}

	return uint(startNum), rangeSize, nil
}
```

This then allows `ParseNetworkVLANList` to be implied, and I've renamed it to `ExpandNetworkVLANList` to indicate what it does.

```go
// expandNetworkVLANList expands VLAN ranges in the form "number" or "start-end" to individual VLAN IDs.
func expandNetworkVLANList(rawVLANValues ...string) ([]uint, error) {
	var networkVLANList []uint
	for i, rawVLAN := range rawVLANValues {
		start, size, err := ParseNetworkVLANRange(rawVLAN)
		if err != nil {
			return nil, fmt.Errorf("Invalid VLAN item %d: %w", i, err)
		}

		// Add each VLAN in the range to the list.
		for vlanID := start; vlanID < start+size; vlanID++ {
			networkVLANList = append(networkVLANList, vlanID)
		}
	}

	return networkVLANList, nil
}
```

And as `expandNetworkVLANList` isn't really related to validation per-se, I would suggest we move this out of `validate` package and have it as part of the `lxd/device/device_utils_network.go` file.

Reply:
I see. That `ParseUint32Range` in the validate package confused me quite a bit (about the code structure), but I reckon I now have a clear picture of what should be where :) Thanks @tomponline

Reply:
Yeah basically it checks if there are 1 or 2 parts to the supplied value, then parses the first value, and if there's a 2nd parses that and figures out the range size, and if not just returns the first value and a range size of 1.